### PR TITLE
Add high-level design of the library.

### PR DESCRIPTION
**Pull request by bogdandrutu**
This is a translation into markdown of the document:
https://docs.google.com/document/d/1qtxoLerxCia3_peeqlqc1-2xyV9ytSclAd7ZNSP2z1w/edit#

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,16 +1,148 @@
...
+This is what we design/describe in the current document.
+
+#### Manually instrumented Frameworks
+We are going to instrument some of the most popular frameworks for each language using the 
+OpenCensus library to allow users to get traces/stats when they use these frameworks.
+
+#### Agents
+Some of the languages may support Agents for instrumentation. Java for example can use byte-code 
+manipulation to provide an agent that automatically instruments a binary. Note: not all the 
+languages support this.
```
Comment (tylerbenson) [inline README.md]: I think this section should be called "automatically instrumented frameworks" and include libraries that might accomplish a similar thing via something like monkey patching.

Another section might separate out instrumentation that is embedded in a framework or runtime.

File: NamespaceAndPackage.md [hunk 2]
```diff
@@ -1,8 +1,7 @@
 # Namespace and Package names
 
 ## Introduction
-This document describes the key package names (namespaces, etc.). This document uses terminology
-(MUST, SHOULD, etc) from [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt).
+This document describes the key package names (namespaces, etc.).
 
 ## Structure
 The top level package name MUST be **opencensus** (for URL model MUST use **io.opencensus**).
```
Comment (rakyll) [inline NamespaceAndPackage.md]: Does this mean we need go.opencensus.io or alike as a Go import path?

File: README.md [hunk 3]
```diff
@@ -1,16 +1,148 @@
...
+The key elements of the Generic Context Support are:
+* Every implementation MUST offer an explicit or implicit generic Context propagation mechanism 
+that allows different sub-contexts to be propagated.
+* Languages that already have this support, like [Go][goContext] or C# (ExecutionContext), MUST 
+use the language supported generic context instead of building their own.
+* For an explicit generic context implementation you can look at the Java [io.grpc.Context][gRPCContext].
+
+
+#### Trace
+Trace component is designed to support distributed tracing (see dapper paper). This component
```
Comment (rakyll) [inline README.md]: give link

File: README.md [hunk 4]
```diff
@@ -1,16 +1,148 @@
...
+#### Trace
+Trace component is designed to support distributed tracing (see dapper paper). This component 
+allows collecting and recording in memory trace data, tracks active requests, and keeps local 
+samples for interesting requests.
+
+The key elements of the API can be broken down as:
+* A Span represents a single operation within a trace. Spans can be nested to form a trace tree. 
+* Libraries must allow users to record tracing events for a span (attributes, annotations, links, 
+etc.).
+* Spans carried in the Generic Context. Libraries MUST provide a means of manipulating the Span in
```
Comment (rakyll) [inline README.md]: Spans are carried.

File: README.md [hunk 5]
```diff
@@ -1,16 +1,148 @@
...
+the context, including getting the current Span, and replacing the current Span.
```
Comment (rakyll) [inline README.md]: Why not a simpler sentence?

Libraries MUST provide a way of getting, manipulating, and replacing the Span in the current context.

File: README.md [hunk 6]
```diff
@@ -1,16 +1,148 @@
...
+* Support dynamically control of the global configuration (e.g. trace sampling rate/probability).
```
Comment (rakyll) [inline README.md]: Is this a MUST?

File: README.md [hunk 7]
```diff
@@ -1,16 +1,148 @@
...
+* A Span represents a single operation within a trace. Spans can be nested to form a trace tree. 
+* Libraries must allow users to record tracing events for a span (attributes, annotations, links, 
+etc.).
+* Spans carried in the Generic Context. Libraries MUST provide a means of manipulating the Span in
+the context, including getting the current Span, and replacing the current Span.
+* Support dynamically control of the global configuration (e.g. trace sampling rate/probability). 
+* Libraries MUST provide a means of dynamically controlling the trace global config at runtime.
+* Libraries SHOULD keep track of active spans and in memory samples based on latency/errors and 
+offer ways to access the data.
+* Because “trace context” must also be propagated between process, library MUST offer the
```
Comment (rakyll) [inline README.md]: Is this trace context or a multi purpose context? 

Because context must also be propagated among processes, library must offer...

File: README.md [hunk 8]
```diff
@@ -1,16 +1,148 @@
...
+* Support dynamically control of the global configuration (e.g. trace sampling rate/probability). 
+* Libraries MUST provide a means of dynamically controlling the trace global config at runtime.
+* Libraries SHOULD keep track of active spans and in memory samples based on latency/errors and 
+offer ways to access the data.
+* Because “trace context” must also be propagated between process, library MUST offer the 
+functionality that allows any transport (e.g RPC, HTTP, etc.) systems to encode/decode the “trace
+context” for placement on the wire.
+
+##### Links
+* Trace API is described [here][TraceAPI]
```
Comment (rakyll) [inline README.md]: These are proper sentences, finish with a full stop.

File: README.md [hunk 9]
```diff
@@ -1,16 +1,148 @@
...
+* Data model is defined [here][TraceDataModel]
+
+#### Tags
+Tags are used by the Stats API to break down measurements by arbitrary metadata set in the 
+current process or propagated from a remote caller (tags are propagated through the Generic 
+Context subsystem inside a process and between process by any transport (e.g RPC, HTTP, etc.)
```
Comment (rakyll) [inline README.md]: and among processes

File: README.md [hunk 10]
```diff
@@ -1,16 +1,148 @@
...
+
+##### Links
+* Trace API is described [here][TraceAPI]
+* Data model is defined [here][TraceDataModel]
+
+#### Tags
+Tags are used by the Stats API to break down measurements by arbitrary metadata set in the 
+current process or propagated from a remote caller (tags are propagated through the Generic 
+Context subsystem inside a process and between process by any transport (e.g RPC, HTTP, etc.) 
+systems).
```
Comment (rakyll) [inline README.md]: Closing the unopened parenthesis.

File: README.md [hunk 11]
```diff
@@ -1,16 +1,148 @@
...
+offer ways to access the data.
+* Because “trace context” must also be propagated between process, library MUST offer the 
+functionality that allows any transport (e.g RPC, HTTP, etc.) systems to encode/decode the “trace
+context” for placement on the wire.
+
+##### Links
+* Trace API is described [here][TraceAPI]
+* Data model is defined [here][TraceDataModel]
+
+#### Tags
```
Comment (rakyll) [inline README.md]: I'd first explain what the tags are, then explain how stats API uses them.

Tags are values propagated through the Generic Context subsystem inside a process and among processes by any transport (e.g RPC, HTTP, etc.). Tags are used by the Stats API to break down measurements by arbitrary metadata set in the  current process or propagated from a remote caller.

File: README.md [hunk 12]
```diff
@@ -1,16 +1,148 @@
...
+* Because tags must also be propagated between process, library MUST offer the functionality that 
+allows RPC systems to encode/decode the set of tags for placement on the wire.
+
+##### Links
+* TODO: Add links to API definition and data model.
+
+#### Stats
+The Stats API is designed to record measurements, dynamically break them down by 
+application-defined tags, and aggregate those measurements in user-defined ways. It is designed 
+to offer multiple types of aggregation (e.g. Distributions) and be efficient (all measurement
```
Comment (rakyll) [inline README.md]: e.g. distributions

File: README.md [hunk 13]
```diff
@@ -1,16 +1,148 @@
...
+
+##### Links
+* TODO: Add links to API definition and data model.
+
+#### Stats
+The Stats API is designed to record measurements, dynamically break them down by 
+application-defined tags, and aggregate those measurements in user-defined ways. It is designed 
+to offer multiple types of aggregation (e.g. Distributions) and be efficient (all measurement 
+processing is done as a background activity); aggregating data enables reducing the overhead of 
+uploading data, while also allowing applications direct access to stats for (e.g.) load-balancing
```
Comment (rakyll) [inline README.md]: for e.g. is strange. Use for example instead.

Comment (rakyll) [reply README.md]: Explain how a load balancer can use this data. You are talking about a specific case I already know but for someone who is reading this for the first time, it is hard to follow.

File: README.md [hunk 14]
```diff
@@ -1,16 +1,148 @@
...
+The key elements the API MUST provide are:
+* Defining what is to be measured (the types of data collected, and their meaning), and how data 
+will be aggregated (e.g. into a distribution, cumulative aggregation vs. deltas, etc.). Libraries
+must offer ways for customer to define Metrics that make sense for their application, and support
+a canonical set for RPC/HTTP systems.
+* Recording data - API's for recording measured values. This data is then broken down by tags 
+carried in the context (e.g. a tag can have a value that describes the current RPC service/method
+name; when RPC latency is recorded, this can be made in a generic call, without having to specify
+the exact method), and aggregated as needed (e.g. a histogram of all latency values).
+* Accessing the aggregated data. This can be filtered by type of data, name of resources used,
```
Comment (rakyll) [inline README.md]: s/type of data/data type
s/name of resources used/resource name

File: NamespaceAndPackage.md (hunk 2 above)
Comment (bogdandrutu) [reply NamespaceAndPackage.md]: Sounds about right. we should chat about this because I am not sure how this go distribution of packages works.

File: README.md (hunk 3 above)
Comment (bogdandrutu) [reply README.md]: Done.

File: README.md (hunk 6 above)
Comment (bogdandrutu) [reply README.md]: Removed this and unify with the next section.

File: README.md (hunk 4 above)
Comment (bogdandrutu) [reply README.md]: Done.

File: README.md (hunk 5 above)
Comment (bogdandrutu) [reply README.md]: Done.

File: README.md (hunk 7 above)
Comment (bogdandrutu) [reply README.md]: Done.

File: README.md (hunk 8 above)
Comment (bogdandrutu) [reply README.md]: Done.

File: README.md (hunk 11 above)
Comment (bogdandrutu) [reply README.md]: Done.

File: README.md (hunk 9 above)
Comment (bogdandrutu) [reply README.md]: Done.

File: README.md (hunk 10 above)
Comment (bogdandrutu) [reply README.md]: Done.

File: README.md (hunk 12 above)
Comment (bogdandrutu) [reply README.md]: Done.

File: README.md (hunk 13 above)
Comment (bogdandrutu) [reply README.md]: Done. for example.

File: README.md (hunk 14 above)
Comment (bogdandrutu) [reply README.md]: Done.

File: README.md (hunk 13 above)
Comment (bogdandrutu) [reply README.md]: Failed to explain in few words so I decided to remove this extra example. we can add more examples later.

File: NamespaceAndPackage.md (hunk 2 above)
Comment (tsloughter) [reply NamespaceAndPackage.md]: More natural to go with `github.com/census-instrumentation/trace`, `github.com/census-instrumentation/stats` isn't it? Or I guess `opencensus.io/trace`?

File: README.md (hunk 1 above)
Comment (bogdandrutu) [reply README.md]: Rephrase it. Not sure I understand about your last sentence. Can you give me more details and I can add it then.

**Comment by bogdandrutu** [comment]
Ping.

File: README.md [hunk 15]
```diff
@@ -1,16 +1,148 @@
...
+specific patterns/idioms. Our goal is that all libraries have a consistent "look and feel".
 
-## Trace Design
-* Trace API is described [here](https://github.com/census-instrumentation/opencensus-specs/blob/master/trace/README.md)
+This repository uses terminology (MUST, SHOULD, etc) from [RFC 2119][RFC2119].
 
-* Data model is defined [here](https://github.com/census-instrumentation/opencensus-proto/blob/master/trace/trace.proto)
+## Overview
+Today, distributed tracing systems and stats collection tend to use unique protocols and 
+specifications for propagating context and sending data to backend processing systems. This
```
Comment (adriancole) [inline README.md]: add a word "sending diagnostic data" Otherwise, your system will be hijacked for data provenance or other workflow tracking reasons :)

File: README.md [hunk 16]
```diff
@@ -1,16 +1,148 @@
...
 
-## Trace Design
-* Trace API is described [here](https://github.com/census-instrumentation/opencensus-specs/blob/master/trace/README.md)
+This repository uses terminology (MUST, SHOULD, etc) from [RFC 2119][RFC2119].
 
-* Data model is defined [here](https://github.com/census-instrumentation/opencensus-proto/blob/master/trace/trace.proto)
+## Overview
+Today, distributed tracing systems and stats collection tend to use unique protocols and 
+specifications for propagating context and sending data to backend processing systems. This 
+is true amongst all the large vendors, and we aim to standardize these APIs and data models.
```
Comment (adriancole) [inline README.md]: "we aim to standardize these APIs and data models."
is our aim to standardize? or is our aim to provide a reliable implementations in service of frameworks and agents. We do that by standardizing APIs and data models.

I think this leads to the next paragraph better

File: README.md [hunk 17]
```diff
@@ -1,16 +1,148 @@
...
 
-## Tags Design
+The OpenCensus library refers to a collection of tools/APIs that are intended to be used primarily
+by cloud and distributed applications for monitoring and performance debugging. OpenCensus is
```
Comment (adriancole) [inline README.md]: defocus from being first. it might not be true

OpenCensus provides a tested set of application performance management (APM) libraries, such as Metrics and Tracing, under a friendly OSS license. We acknowlege the polyglot nature of modern applications, and provide implementations in all main programming languages....

File: README.md [hunk 18]
```diff
@@ -1,16 +1,148 @@
...
 
-## Tags Design
+The OpenCensus library refers to a collection of tools/APIs that are intended to be used primarily
+by cloud and distributed applications for monitoring and performance debugging. OpenCensus is 
+intended to be the first open-source application performance management (APM) library available 
+in all the main programming languages including C/C++, Java, Go, Ruby, PHP, Python, C#, Node.js,
+Objective-C and Erlang.
 
-## Stats Design
+## Ecosystem Design
```
Comment (adriancole) [inline README.md]: Integration design

File: README.md [hunk 19]
```diff
@@ -1,16 +1,148 @@
...
 
-## Supported Encodings
+![alt text][EcosystemLayers]
+
+### Layers
+
+#### Service Exporters
+Each backend service MUST implement this API to export data to their services.
```
Comment (adriancole) [inline README.md]: Backends can choose to accept data in a supported format, or implement this api to accept in theirs

File: README.md [hunk 20]
```diff
@@ -1,16 +1,148 @@
...
+
+#### OpenCensus Library
+This is what we design/describe in the current document.
+
+#### Manually instrumented Frameworks
+We are going to instrument some of the most popular frameworks for each language using the
```
Comment (adriancole) [inline README.md]: For example, we a configuration approach to enable tracing and metrics in gRPC applications

File: README.md [hunk 21]
```diff
@@ -1,16 +1,148 @@
...
+
+#### OpenCensus Library
+This is what we design/describe in the current document.
+
+#### Manually instrumented Frameworks
+We are going to instrument some of the most popular frameworks for each language using the 
+OpenCensus library to allow users to get traces/stats when they use these frameworks.
+
+#### Tools for automatic instrumentation
+Some of the languages may support libraries for automatic instrumentation. Java for example can use
```
Comment (adriancole) [inline README.md]: Some users want to "turn on" metrics or tracing, automatically via agents or frameworks provided to them by a vendor. Java for example... Automatic tracing is a key driver for our library design decisions.

File: README.md [hunk 22]
```diff
@@ -1,16 +1,148 @@
...
+### Components
+This section focuses on the important components that each OpenCensus library must have to 
+support all the functionalities.
+
+Here is a layering structure of the proposed OpenCensus library:
+
+![alt text][LibraryComponents]
+
+
+#### Generic Context
```
Comment (adriancole) [inline README.md]: Propagated Context (hits much harder.. .generic is a filler word)

File: README.md [hunk 23]
```diff
@@ -1,16 +1,148 @@
...
+Some of the features like tracing (distributed tracing) and tagging (possibly others) need a way 
+to propagate a specific context (trace, tags) in-process (possibly between threads) and function
+calls.
+
+The key elements of the Generic Context Support are:
+* Every implementation MUST offer an explicit or implicit generic Context propagation mechanism 
+that allows different sub-contexts to be propagated.
+* Languages that already have this support, like [Go][goContext] or C# (ExecutionContext), MUST 
+use the language supported generic context instead of building their own.
+* For an explicit generic context implementation you can look at the Java [io.grpc.Context][gRPCContext].
```
[the remark this answers is not in the document]
Comment (adriancole) [reply README.md]: explicit and gRPC context are sortof a mixed thing, as grpc context includes hooks to instrument executors

File: README.md [hunk 24]
```diff
@@ -1,16 +1,148 @@
...
+
+
+#### Trace
+Trace component is designed to support distributed tracing (see [dapper paper][DapperPaper]). This
```
Comment (adriancole) [inline README.md]: second sentence ...

Census allows functionality beyond data collection and export, for example, tracking active requests...

File: README.md [hunk 25]
```diff
@@ -1,16 +1,148 @@
...
+* A Span represents a single operation within a trace. Spans can be nested to form a trace tree. 
+* Libraries must allow users to record tracing events for a span (attributes, annotations, links, 
+etc.).
+* Spans are carried in the Generic Context. Libraries MUST provide a way of getting, manipulating,
+and replacing the Span in the current context.
+* Libraries MUST provide a means of dynamically controlling the trace global config at runtime (e
+.g. trace sampling rate/probability).
+* Libraries SHOULD keep track of active spans and in memory samples based on latency/errors and 
+offer ways to access the data.
+* Because context must also be propagated among processes, library MUST offer the functionality
```
Comment (adriancole) [inline README.md]: "propagated among processes" "propagated across processes"

File: README.md [hunk 26]
```diff
@@ -1,16 +1,148 @@
...
+placement on the wire.
+
+##### Links
+* Trace API is described [here][TraceAPI].
+* Data model is defined [here][TraceDataModel].
+
+#### Tags
+Tags are values propagated through the Generic Context subsystem inside a process and among 
+processes by any transport (e.g RPC, HTTP, etc.). Tags are used by the Stats component to break 
+down  measurements by arbitrary metadata set in the current process or propagated from a remote
```
Comment (adriancole) [inline README.md]: tags are only used by stats? If not. add "for example, " in front of the "Tags are used" sentence

File: README.md [hunk 27]
```diff
@@ -1,16 +1,148 @@
...
+The key elements the API MUST provide are:
+* Defining what is to be measured (the types of data collected, and their meaning), and how data 
+will be aggregated (e.g. into a distribution, cumulative aggregation vs. deltas, etc.). Libraries
+must offer ways for customer to define Metrics that make sense for their application, and support
+a canonical set for RPC/HTTP systems.
+* Recording data - API's for recording measured values. This data is then broken down by tags 
+carried in the context (e.g. a tag can have a value that describes the current RPC service/method
+name; when RPC latency is recorded, this can be made in a generic call, without having to specify
+the exact method), and aggregated as needed (e.g. a histogram of all latency values).
+* Accessing the aggregated data. This can be filtered by data type, resource name, etc. This
```
Comment (adriancole) [inline README.md]: this is a big feature with little text :) maybe put an example of aggregation or view

File: README.md [hunk 28]
```diff
@@ -1,16 +1,148 @@
...
+component allows collecting and recording in memory trace data, tracks active requests, and keeps
+local samples for interesting requests.
+
+The key elements of the API can be broken down as:
+* A Span represents a single operation within a trace. Spans can be nested to form a trace tree. 
+* Libraries must allow users to record tracing events for a span (attributes, annotations, links, 
+etc.).
+* Spans are carried in the Generic Context. Libraries MUST provide a way of getting, manipulating,
+and replacing the Span in the current context.
+* Libraries MUST provide a means of dynamically controlling the trace global config at runtime (e
```
Comment (chingor13) [inline README.md]: Is this truly a MUST? I don't think it will be feasible for PHP to implement this.

Comment (jcchavezs) [reply README.md]: :+1: I'd reword it as: *Libraries SHOULD provide a means of dynamically controlling the trace global config at runtime (immediately or between requests).

Comment (chingor13) [reply README.md]: I think I misunderstood this line. It must be configurable at runtime, but not necessarily changable while running.

File: README.md [hunk 29]
```diff
@@ -1,16 +1,148 @@
...
+name; when RPC latency is recorded, this can be made in a generic call, without having to specify
+the exact method), and aggregated as needed (e.g. a histogram of all latency values).
+* Accessing the aggregated data. This can be filtered by data type, resource name, etc. This 
+allows applications to easily get access to their own data in-process.
+
+##### Links
+* TODO: Add links to API definition and data model.
+
+### Supported propagation formats
+* Library MUST support the [TraceContext][TraceContextSpecs] format for Trace and Tags components.
```
Comment (liyanhui1228) [inline README.md]: What about Cloud Trace format propagation? Do we need to add it here?

File: README.md (hunk 15 above)
Comment (bogdandrutu) [reply README.md]: Done.

File: README.md (hunk 16 above)
Comment (bogdandrutu) [reply README.md]: Done.

File: README.md (hunk 17 above)
Comment (bogdandrutu) [reply README.md]: Done.

File: README.md (hunk 18 above)
Comment (bogdandrutu) [reply README.md]: I like more ecosystem sorry.

File: README.md (hunk 19 above)
Comment (bogdandrutu) [reply README.md]: I would not suggest this here. I relaxed to SHOULD so the option is there but don't want to have that as a first option.

File: README.md (hunk 20 above)
Comment (bogdandrutu) [reply README.md]: Don't understand what you want to say here.

File: README.md (hunk 22 above)
Comment (bogdandrutu) [reply README.md]: I changed it as just Context, if you care about calling "propagated" we can change it later.

File: README.md (hunk 24 above)
Comment (bogdandrutu) [reply README.md]: Done.

File: README.md (hunk 25 above)
Comment (bogdandrutu) [reply README.md]: Done.

File: README.md (hunk 26 above)
Comment (bogdandrutu) [reply README.md]: Done.

File: README.md (hunk 27 above)
Comment (bogdandrutu) [reply README.md]: We can rewrite some of this, but I think is more important to ship this than focusing on these things.

File: README.md (hunk 24 above)
Comment (bogdandrutu) [reply README.md]: Done.

File: NamespaceAndPackage.md (hunk 2 above)
Comment (bogdandrutu) [reply NamespaceAndPackage.md]: Filed an issue for this:
https://github.com/census-instrumentation/opencensus-specs/issues/15

File: README.md (hunk 28 above)
Comment (bogdandrutu) [reply README.md]: Why is it hard to do it at runtime (change)? I relaxed it to SHOULD but I would like to understand.
